clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -424,19 +457,27 @@ public ConfigDef withClientSaslSupport() {
             String joined = Utils.join(undefinedConfigKeys, ",");
             throw new ConfigException("Some configurations in are referred in the dependents, but not defined: " + joined);
         }
-        // parse all known keys
+        // Handle all known ConfigKey objects.
         Map<String, Object> values = new HashMap<>();
         for (ConfigKey key : configKeys.values()) {
-            Object value;
-            // props map contains setting - assign ConfigKey value
+            Object value = key.defaultValue;
             if (props.containsKey(key.name)) {
                 value = parseType(key.name, props.get(key.name), key.type);
-                // props map doesn't contain setting, the key is required because no default value specified - its an error
-            } else if (key.defaultValue == NO_DEFAULT_VALUE) {
-                throw new ConfigException("Missing required configuration \"" + key.name + "\" which has no default value.");
             } else {
-                // otherwise assign setting its default value
-                value = key.defaultValue;
```

Comment:
The object is initialized to `key.defaultValue`, so we don't need another clause at the end.  But it might be easier to understand with that at the end, so I'll re-add it.